robust-software.tex

```diff
@@ -52,55 +56,56 @@ \section*{Introduction}
+a graduated student's code to analyze their data.
+Often, that software will be undocumented and work in unexpected
+ways (if it works at all).  It will often rely on nonexistent paths or resources,
+be tuned for a single dataset,
+or simply be an older version than was used in published papers.
+Its new user is often faced with two unpalatable options:
+hack the existing code to make it work, or start over.
+
+The root cause of this problem is that most research software
+isn't \emph{robust}. The difference between running and being
 robust is the difference between ``works for me on my machine'' and
 ``works for other people on a cluster I've never used''. In particular,
 robust software:
 
 \begin{itemize}
 \item
-  Works for users other than the original author
+  works for users other than the original author;
 \item
-  Is kept under version control.
+  is kept under version control;
 \item
-  Can be installed on more than one computer or account with relative ease
+  can be installed on more than one computer or account with relative ease;
 \item
-  Has well-defined input and output formats
+  has well-defined input and output formats;
 \item
-  Has documentation that describes what its dependencies are, how to
-  install it, and what the options are.
+  has documentation that describes what its dependencies are, how to
+  install it, and what the options are; and
 \item
-  Comes with enough tests to show that it actually runs.
+  comes with enough tests to show that it actually runs.
 \end{itemize}
 
-These rules need not be applied to \emph{every} coding effort. The vast majority of
-code produced in the marathon of a graduate thesis is used once to answer a
-specific question related to a specific dataset. Scientific endeavours often
-require speed and rapid iteration over a question, not thoughtful software
-engineering and thorough documentation. A README is of no use if you are the
-sole user of your code. As the saying goes, not everything worth doing is worth
-doing right away. However, once your script is dusted off and run three or four
-times for slightly different purposes, when it comes time to publish your
-methods, to leave a project, or to pass it on to someone else, it may be time
-to apply ten simple rules for robust software.
-
-These simple rules are also necessary steps toward creating a reusable library that
-can be shared through a site like CPAN or CRAN. They can apply to both
-closed-source and open-source software.  They are generic and can be applied to all
-languages, libraries, packages, documentation styles, and
-operating systems.  Whether the aim is as simple as sharing the code
-with collaborators or as complex as using the software in a production
-analysis environment, increasing the robustness of your software
-decreases headaches all around.
+These rules need not be applied to \emph{every} coding effort:
+as the saying goes, not everything worth doing is worth doing right (or right away).
```

Comment:
Well, my dad used to say it...  and there's something very similar in 
"The Soul of a New Machine" (quoted on 
https://en.wikipedia.org/wiki/The_Soul_of_a_New_Machine).